### PR TITLE
Add phasor_from_signal function

### PR DESCRIPTION
**Pull request by schutyb**
## Description

... This PR adds a phasor module in phasorpy that compute the phasor transform. 

## Release note

1. Add phasorpy.phasor 
2. Add a test file for the phasor function 

```release-note
...
```

## Checklist

- [ ] The pull request title, summary, and description are concise.
- [ ] Related issues are linked in the description.
- [ ] New dependencies are explained.
- [ ] The source code and documentation can be distributed under the [MIT license](https://www.phasorpy.org/stable/license.html).
- [ ] The source code adheres to [code standards](https://www.phasorpy.org/stable/contributing.html#code-standards).
- [ ] New classes, functions, and features are thoroughly [tested](https://www.phasorpy.org/stable/contributing.html#tests).
- [ ] New, user-facing classes, functions, and features are [documented](https://www.phasorpy.org/stable/contributing.html#documentation).
- [ ] New features are covered in tutorials.
- [ ] No files other than source code, documentation, and project settings are added to the repository.

### CODE REVIEW

**Comment by cgohlke** [comment]
I already reviewed #19, which looks identical to this PR.

**Comment by schutyb** [comment]
> I already reviewed #19, which looks identical to this PR.

I just deleted #19 because I did not followed the doc to fork the repository before, so I restart by doing it.
In #19 I didn´t see any corrections.

**Comment by cgohlke** [comment]
> In https://github.com/phasorpy/phasorpy/pull/19 I didn´t see any corrections.

Strange. I'll copy the comments...

File: src/phasorpy/phasor.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
+"""
+============================================
+Phasor functions, module: "phasorpy.phasor"
+============================================
+
```
Comment (cgohlke) [inline src/phasorpy/phasor.py:5]: The heading is already defined in the documentation phasor.rst

File: src/phasorpy/phasor.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+This module provides a set of function to 
+compute phasor transform based on the fast 
+fourier transform algorithm from numpy.fft. 
+There are general tools common to FLIM and 
+HSI data, that helps when doing phasor 
+analysis. 
+
```
Comment (cgohlke) [inline src/phasorpy/phasor.py:12]: Use [numpy](https://numpydoc.readthedocs.io/en/latest/format.html) and [Python](https://peps.python.org/pep-0257/) docstring conventions, starting with a single line summary. Something like

```
Compute phasor coordinates from fluorescence lifetime and hyperspectral signals.

The ``phasorpy.phasor`` module provides a function to calculate 
phasor coordinates (average intensity, real, and imaginary parts) 
from fluorescence lifetime and hyperspectral imaging using Fourier transform...
```

File: src/phasorpy/phasor.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    Returns
+    -------
+        dc: nd array or tuple.
+            Contains the average of the 0-axis
+        g: nd array or tuple.
+            Contains the real part of the fft transform.
+        s: nd array or tuple.
+            Contains the imaginary part of the fft transform.
+
+    """
```
Comment (cgohlke) [inline src/phasorpy/phasor.py:52]: The function should accept array-like, not just image stacks. Ideally it should work with any kind and any dimensional array input, including list, tuples, numpy arrays, xarrays, and objects implementing the buffer protocol.

As discussed in our last call, the ``axis`` over which FFT is performed should be an argument. It could be determined from the xarray metadata.

How about renaming ``imstack`` to ``data`` or ``signal`` and make it an positional-only argument? ``axis`` and ``harmonic`` should be keyword-only arguments?

Why/when can the input array be complex?

Document that the input signal must be ordered and evenly-spaced/uniform. Some frequency-domain or spectral signals might not fulfill these requirements.

fft -> FFT

fourier -> Fourier

Non-numpy FFT functions may be more efficient or more general, e.g. [scipy.fft.fft](https://docs.scipy.org/doc/scipy/reference/generated/scipy.fft.fft.html), [mkl-fft](https://pypi.org/project/mkl-fft/), or [CuPy](https://docs.cupy.dev/en/stable/user_guide/fft.html). Maybe, allow the user to pass a FFT function?

FFTs may consume more memory than necessary, computing and storing all harmonics. First harmonics may be more efficiently calculated "manually" by multiplying with sin/cos terms. This could be added later.

Follow numpy and Python docstring standards.

File: src/phasorpy/phasor.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    if imstack.any():
```
Comment (cgohlke) [inline src/phasorpy/phasor.py:53]: Why is this necessary? The signal will almost never be all empty.

File: src/phasorpy/phasor.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        data = np.fft.fft(imstack, axis=0, norm='ortho')
```
Comment (cgohlke) [inline src/phasorpy/phasor.py:54]: Should this not be `'forward'`, that is normalized by `1/n`?

Make `axis` an argument.

File: src/phasorpy/phasor.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        dc = data[0].real
+        dc = np.where(
+            dc != 0, dc, np.NaN
+        )  # Set NaN where there is a 0 division
```
Comment (cgohlke) [inline src/phasorpy/phasor.py:58]: Isn't it more efficient to let the division handle the errors?

File: src/phasorpy/phasor.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        g = data[harmonic].real
+        g /= dc
+        s = data[harmonic].imag
+        s /= dc
```
Comment (cgohlke) [inline src/phasorpy/phasor.py:62]: I think `g`, `s` , and likely `dc` are just views into the array returned by the FFT. That means the large array is kept in memory even though only a small part of it is used. It is probably better to return copies of `dc`, `g`, and `s`.

I prefer to use `real` and `imag` as variable names rather than single letter `g` and `s`.

File: tests/test_phasor.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    assert s.shape == (data.shape[1], data.shape[2])
+    assert numpy.all(dc != 0)
+
+
+def test_empty_phasor():
+    """Test empty image phasor calculation error"""
+    empty_data = numpy.zeros((30, 512, 512))
+    with pytest.raises(ValueError):
+        phasor(empty_data)
+
```
Comment (cgohlke) [inline tests/test_phasor.py:55]: The `harmonic` argument is not tested. 

Computed values are not compared to known values.

**Comment by schutyb** [comment]
> > In #19 I didn´t see any corrections.
> 
> Strange. I'll copy the comments...

Great I´m doing some changes and update it soon, thanks

**Comment by cgohlke** [comment]
This PR conflicts with the recent additions to the main branch. Can you update your [phasor-function](https://github.com/schutyb/phasorpy/tree/phasor-function) branch with the latest [main](https://github.com/phasorpy/phasorpy) branch?

**Comment by cgohlke** [comment]
For reference, here's the preliminary `phasor_from_signal` function I just removed from #32:
```Python
def phasor_from_signal(
    signal: ArrayLike,
    /,
    *,
    axis: int = 0,
    harmonic: int = 1,
) -> tuple[NDArray[Any], NDArray[Any], NDArray[Any]]:
    """Return phasor coordinates from signal using FFT.

    Parameters
    ----------
    signal : array_like
        Frequency-domain, time-domain, or hyperspectral data.
        A minimum of three uniformly spaced samples are required along `axis`.
    axis : int, optional
        Axis over which to compute phasor coordinates.
        The default is the first axis.
    harmonic: int, optional
        Index in FFT result containing real and imag components. Must be >= 1.
        The default is the first harmonic.

    Returns
    -------
    mean : ndarray
        Average of signal along axis.
    real : ndarray
        Real component of phasor coordinates along axis.
    imag : ndarray
        Imaginary component of phasor coordinates along axis.

    Raises
    ------
    ValueError
        The `signal` has less than three samples along `axis`.
        The `sample_phase` size does not match number of samples along `axis`.
        The `harmonic` is smaller than 1.

    Examples
    --------
    Calculate phasor coordinates of a phase-shifted sinusoidal signal:

    >>> sample_phase = numpy.linspace(0, 2 * math.pi, 5, endpoint=False)
    >>> signal = 1.1 * (
    ...     numpy.cos(sample_phase - 0.78539816) * 2 *  0.70710678 + 1
    ... )
    >>> phasor_from_signal(signal)  # doctest: +NUMBER
    (1.1, 0.5, 0.5)

    The sinusoidal signal does not have a second harmonic component:

    >>> phasor_from_signal(signal, harmonic=2)  # doctest: +NUMBER
    (1.1, 0.0, 0.0)

    """
    signal = numpy.array(signal, copy=False, ndmin=1)
    samples = numpy.size(signal, axis)
    if samples < 3:
        raise ValueError(f'not enough {samples=} along {axis=}')
    if harmonic < 1:
        raise ValueError(f'{harmonic=} < 1')
    fft: NDArray = numpy.fft.fft(signal, axis=axis, norm='forward')
    mean = fft.take(0, axis=axis).real.copy()
    fft = fft.take(harmonic, axis=axis)
    real = fft.real.copy()
    real /= mean
    imag = fft.imag.copy()
    imag /= mean
    imag *= -1
    return mean, real, imag


def test_phasor_from_signal():
    """Test `phasor_from_signal` function."""
    sample_phase = numpy.linspace(0, 2 * math.pi, 5, endpoint=False)
    signal = 1.1 * (numpy.cos(sample_phase - 0.78539816) * 2 * 0.70710678 + 1)
    assert_allclose(phasor_from_signal(signal), (1.1, 0.5, 0.5), atol=1e-6)
    # the sinusoidal signal does not have a second harmonic component
    assert_allclose(
        phasor_from_signal(signal, harmonic=2), (1.1, 0.0, 0.0), atol=1e-6
    )
    with pytest.raises(ValueError):
        phasor_from_signal(signal[:2])
    with pytest.raises(ValueError):
        phasor_from_signal(signal, harmonic=0)
    # TODO: more tests (dimensions, axis, types, NaN, ...)
```

**Comment by schutyb** [comment]
> This PR conflicts with the recent additions to the main branch. Can you update your [phasor-function](https://github.com/schutyb/phasorpy/tree/phasor-function) branch with the latest [main](https://github.com/phasorpy/phasorpy) branch?

Should I checkout my branch and then merge main branch?
 Then add the changes in phasor.py and test_phasor.py, and afterthat commit it?

**Comment by schutyb** [comment]
> For reference, here's the preliminary `phasor_from_signal` function I just removed from #32:
> 
> ```python
> def phasor_from_signal(
>     signal: ArrayLike,
>     /,
>     *,
>     axis: int = 0,
>     harmonic: int = 1,
> ) -> tuple[NDArray[Any], NDArray[Any], NDArray[Any]]:
>     """Return phasor coordinates from signal using FFT.
> 
>     Parameters
>     ----------
>     signal : array_like
>         Frequency-domain, time-domain, or hyperspectral data.
>         A minimum of three uniformly spaced samples are required along `axis`.
>     axis : int, optional
>         Axis over which to compute phasor coordinates.
>         The default is the first axis.
>     harmonic: int, optional
>         Index in FFT result containing real and imag components. Must be >= 1.
>         The default is the first harmonic.
> 
>     Returns
>     -------
>     mean : ndarray
>         Average of signal along axis.
>     real : ndarray
>         Real component of phasor coordinates along axis.
>     imag : ndarray
>         Imaginary component of phasor coordinates along axis.
> 
>     Raises
>     ------
>     ValueError
>         The `signal` has less than three samples along `axis`.
>         The `sample_phase` size does not match number of samples along `axis`.
>         The `harmonic` is smaller than 1.
> 
>     Examples
>     --------
>     Calculate phasor coordinates of a phase-shifted sinusoidal signal:
> 
>     >>> sample_phase = numpy.linspace(0, 2 * math.pi, 5, endpoint=False)
>     >>> signal = 1.1 * (
>     ...     numpy.cos(sample_phase - 0.78539816) * 2 *  0.70710678 + 1
>     ... )
>     >>> phasor_from_signal(signal)  # doctest: +NUMBER
>     (1.1, 0.5, 0.5)
> 
>     The sinusoidal signal does not have a second harmonic component:
> 
>     >>> phasor_from_signal(signal, harmonic=2)  # doctest: +NUMBER
>     (1.1, 0.0, 0.0)
> 
>     """
>     signal = numpy.array(signal, copy=False, ndmin=1)
>     samples = numpy.size(signal, axis)
>     if samples < 3:
>         raise ValueError(f'not enough {samples=} along {axis=}')
>     if harmonic < 1:
>         raise ValueError(f'{harmonic=} < 1')
>     fft: NDArray = numpy.fft.fft(signal, axis=axis, norm='forward')
>     mean = fft.take(0, axis=axis).real.copy()
>     fft = fft.take(harmonic, axis=axis)
>     real = fft.real.copy()
>     real /= mean
>     imag = fft.imag.copy()
>     imag /= mean
>     imag *= -1
>     return mean, real, imag
> 
> 
> def test_phasor_from_signal():
>     """Test `phasor_from_signal` function."""
>     sample_phase = numpy.linspace(0, 2 * math.pi, 5, endpoint=False)
>     signal = 1.1 * (numpy.cos(sample_phase - 0.78539816) * 2 * 0.70710678 + 1)
>     assert_allclose(phasor_from_signal(signal), (1.1, 0.5, 0.5), atol=1e-6)
>     # the sinusoidal signal does not have a second harmonic component
>     assert_allclose(
>         phasor_from_signal(signal, harmonic=2), (1.1, 0.0, 0.0), atol=1e-6
>     )
>     with pytest.raises(ValueError):
>         phasor_from_signal(signal[:2])
>     with pytest.raises(ValueError):
>         phasor_from_signal(signal, harmonic=0)
>     # TODO: more tests (dimensions, axis, types, NaN, ...)
> ```

Ok, I passed fft to my phasor_from_signal so the user can pass any fft module, Should I remove it? I will add more RaiseValue.

**Comment by cgohlke** [comment]
> Ok, I passed fft to my phasor_from_signal so the user can pass any fft module, Should I remove it? 

That's not what I meant. The code I pasted was just to preserve what I removed from #32. Maybe some things can be reused here, maybe not.

Let's discuss the fft callback function when the PR is ready to review.

**Comment by cgohlke** [comment]
> Should I checkout my branch and then merge main branch?
> Then add the changes in phasor.py and test_phasor.py, and afterthat commit it?

Did you try what GitHub suggests in the two links under "This branch has conflicts that must be resolved"?

**Comment by schutyb** [comment]
> > Should I checkout my branch and then merge main branch?
> > Then add the changes in phasor.py and test_phasor.py, and afterthat commit it?
> 
> Did you try what GitHub suggests in the two links under "This branch has conflicts that must be resolved"?

I did

git checkout -b schutyb-phasor-function main
git pull https://github.com/schutyb/phasorpy.git phasor-function

and then 

git checkout main
git merge --no-ff schutyb-phasor-function
git push origin main 

But I don´t see where any changes

**Comment by cgohlke** [comment]
I don't know in what state your local repository is. From https://github.com/phasorpy/phasorpy/network it looks like the recent changes in this repository were never pulled into your branch. Did you manually copy the changes? Also your main seems to have some extra commits. Maybe you can meet with @bruno-pannunzio in person to resolve the conflicts.

**Comment by schutyb** [comment]
> I don't know in what state your local repository is. From https://github.com/phasorpy/phasorpy/network it looks like the recent changes in this repository were never pulled into your branch. Did you manually copy the changes? Also your main seems to have some extra commits. Maybe you can meet with @bruno-pannunzio in person to resolve the conflicts.

@cgohlke We resolved the conflict I got in my branch but some test are not running and we don´t understand why. Can you have a look and let us know?

**Comment by cgohlke** [comment]
Looks like yesterday's update of the [pooch library](https://pypi.org/project/pooch/) broke things for us. I'll try to fix that. In the meantime, use ``pooch==1.8.1`` in the dependencies.

**Comment by schutyb** [comment]
> Looks like yesterday's update of the [pooch library](https://pypi.org/project/pooch/) broke things for us. I'll try to fix that. In the meantime, use `pooch==1.8.0` in the dependencies.

It worked now, but cibuildwheel didn´t. What does cibuildwheel do? Is it use to facilitate the packeages intalling?

**Comment by cgohlke** [comment]
> some test are not running

Fixed in #34.

**Comment by cgohlke** [comment]
Given the huge performance and memory advantage of ``phasor_from_signal_f1`` compared to the ``numpy.fft`` reference implementation (see https://github.com/phasorpy/phasorpy/pull/32#issuecomment-1932748126), it might not be worth perusing the FFT approach. Does scipy.fft have the same issue? ``phasor_from_signal_f1`` could be expanded to calculate harmonics instead...
